ark/jsonschema/__tests__/object.test.ts

```diff
@@ -0,0 +1,148 @@
+import { attest, contextualize } from "@ark/attest"
+import { parseJsonSchema } from "@ark/jsonschema"
+
+contextualize(() => {
+	it("type object", () => {
+		const t = parseJsonSchema({ type: "object" })
+		attest(t.json).snap({ domain: "object" })
+	})
+
+	it("maxProperties", () => {
+		const tMaxProperties = parseJsonSchema({
+			type: "object",
+			maxProperties: 1
+		})
+		attest(tMaxProperties.json).snap({
+			domain: "object",
+			predicate: ["$ark.jsonSchemaObjectMaxPropertiesValidator"]
+		})
+		attest(tMaxProperties.allows({})).equals(true)
+		attest(tMaxProperties.allows({ foo: 1 })).equals(true)
+		attest(tMaxProperties.allows({ foo: 1, bar: 2 })).equals(false)
+		attest(tMaxProperties.allows({ foo: 1, bar: 2, baz: 3 })).equals(false)
+	})
+
+	it("minProperties", () => {
+		const tMinProperties = parseJsonSchema({
+			type: "object",
+			minProperties: 2
+		})
+		attest(tMinProperties.json).snap({
+			domain: "object",
+			predicate: ["$ark.jsonSchemaObjectMinPropertiesValidator"]
+		})
+		attest(tMinProperties.allows({})).equals(false)
+		attest(tMinProperties.allows({ foo: 1 })).equals(false)
+		attest(tMinProperties.allows({ foo: 1, bar: 2 })).equals(true)
+		attest(tMinProperties.allows({ foo: 1, bar: 2, baz: 3 })).equals(true)
+	})
+
+	it("properties & required", () => {
+		const tRequired = parseJsonSchema({
+			type: "object",
+			properties: {
+				foo: { type: "string" },
+				bar: { type: "number" }
+			},
+			required: ["foo"]
+		})
+		attest(tRequired.json).snap({
+			domain: "object",
+			required: [{ key: "foo", value: "string" }],
+			optional: [{ key: "bar", value: "number" }]
+		})
+
+		attest(() => parseJsonSchema({ type: "object", required: ["foo"] })).throws(
+			"TraversalError: must be a valid object JSON Schema (was an object JSON Schema with 'required' array but no 'properties' object)"
+		)
+		attest(() =>
+			parseJsonSchema({
+				type: "object",
+				properties: { foo: { type: "string" } },
+				required: ["bar"]
+			})
+		).throws(
+			`TraversalError: required must be a key from the 'properties' object (one of ["foo"]) (was bar)`
+		)
+		attest(() =>
+			parseJsonSchema({
+				type: "object",
+				properties: { foo: { type: "string" } },
+				required: ["foo", "foo"]
+			})
+		).throws(
+			`TraversalError: required must be an array of unique strings (was an array with the following duplicates: [{"element":"foo","indices":[1]}])`
```

Comment:
nit: should we include the first index here as well for clarity in case the list is longer?

Reply:
Yeah, I agree the first index should be included. Updated in cc44db7 (and even fixed a bug in the process!)